enn_zoo/enn_zoo/microrts/__init__.py

```diff
@@ -0,0 +1,360 @@
+            os.path.expanduser(self.microrts_path),
+            self.map_path,
+            self.ai2s[0](self.real_utt),
+            self.real_utt,
+            self.partial_obs,
+        )
+        # get the unit type table
+        self.utt = json.loads(str(self.client.sendUTT()))
+
+    @classmethod
+    def obs_space(cls) -> ObsSpace:
+        return ObsSpace(
+            {
+                "Resource": Entity(["x", "y"]),
+                "Base": Entity(["x", "y"]),
+                "Barracks": Entity(["x", "y"]),
+                "Worker": Entity(["x", "y"]),
+                "Light": Entity(["x", "y"]),
+                "Heavy": Entity(["x", "y"]),
+                "Ranged": Entity(["x", "y"]),
+            }
+        )
+
+    @classmethod
+    def action_space(cls) -> Dict[str, ActionSpace]:
+        return {
+            "unit_action": CategoricalActionSpace(
+                choices=[
+                    "move_up",
+                    "move_right",
+                    "move_down",
+                    "move_left",
+                    "harvest_up",
+                    "harvest_right",
+                    "harvest_down",
+                    "harvest_left",
+                    "return_up",
+                    "return_right",
+                    "return_down",
+                    "return_left",
+                    "produce_base_up",
+                    "produce_base_right",
+                    "produce_base_down",
+                    "produce_base_left",
+                    "produce_barrack_up",
+                    "produce_barrack_right",
+                    "produce_barrack_down",
+                    "produce_barrack_left",
+                ]
+                + [
+                    f"attack_location_{i}" for i in range(49)
+                ],  # the attack trange is a 7x7 relative grid
+            ),
+            "base_action": CategoricalActionSpace(
+                choices=[
+                    "produce_worker_up",
+                    "produce_worker_right",
+                    "produce_worker_down",
+                    "produce_worker_left",
+                ],
+            ),
+            "barrack_action": CategoricalActionSpace(
+                choices=[
+                    "produce_light_up",
+                    "produce_light_right",
+                    "produce_light_down",
+                    "produce_light_left",
+                    "produce_heavy_up",
+                    "produce_heavy_right",
+                    "produce_heavy_down",
+                    "produce_heavy_left",
+                    "produce_ranged_up",
+                    "produce_ranged_right",
+                    "produce_ranged_down",
+                    "produce_ranged_left",
+                ],
+            ),
+        }
+
+    def reset(self) -> Observation:
+        self.step = 0
+        self.returns = np.zeros(len(self.rfs))
+
+        response = self.client.reset(0)
+        # self.client.render(False)
```

Comment:
Add an env parameter to enable/disable rendering?

Reply:
Better: Included video recording by implementing the `render` function. See https://wandb.ai/entity-neural-network/enn-ppo/runs/18ca9rdm?workspace=user-costa-huang